widget/embedded/src/store/slices/data.ts

```diff
@@ -81,23 +82,34 @@ export const createDataSlice: StateCreator<
     }
 
     const config = get().config;
-    const supportedTokensFromConfig =
+    const supportedTokensConfig =
       (options.type === 'source' ? config.from?.tokens : config.to?.tokens) ??
-      [];
+      {};
     const blockchains = get().blockchains({
       type: options.type,
     });
 
     const list = tokensFromState
       .filter((token) => {
-        // If there is a list of tokens in config, we only keep them.
-        if (
-          supportedTokensFromConfig.length > 0 &&
-          !supportedTokensFromConfig.some((asset) => {
-            return tokensAreEqual(asset, token);
-          })
-        ) {
-          return false;
+        if (supportedTokensConfig) {
+          const hasToken = handleHasToken(token, supportedTokensConfig);
+          if (Array.isArray(supportedTokensConfig)) {
+            if (supportedTokensConfig.length > 0 && !hasToken) {
+              return false;
+            }
+          } else if (supportedTokensConfig[token.blockchain]) {
+            const isExclude = isExcluded(
+              supportedTokensConfig,
+              token.blockchain
+            );
+            if (!isExclude && !hasToken) {
+              // If there is a list of tokens in the configuration and isExcluded is false, we only keep them.
+              return false;
+            } else if (isExclude && hasToken) {
+              // If there is a list of tokens in the configuration and isExcluded is true, we don't keep them.
+              return false;
+            }
+          }
```

Comment:
extract a function like `isTokenInConfig` and change it like this:
```ts
const list = tokensFromState.filter((token) => { isTokenInConfig(supportedTokensConfig, token)  })
```
Also use better variable names there.

Reply:
fixed